### DISCUSSION TITLE
Include maxial values in overlay texts

### DISCUSSION THREAD

**Discussion by KosmoJoe** [time4tea/gopro-dashboard-overlay#113] [General]
Hi,

first of all thank you for this extremely usefull and flexible tool to make nice overlays to go-pro videos. It is really fun to play with it.

I started with adapting the example xml files for my personal perference. I would like to realize an overlay with a text saying:
"Distance (m)     340/1523"
where the 340 is the distance from the start and the 1523 is the total distance. While to former I was able to realise with the help of your documentation using the odo metric, the latter I dont't know how to achieve. Is there the possiblity to include text with "MAX(odo)"? This would also be interesting for speed/height/slope.

Bests
KosmoJoe

**Comment by time4tea** [comment]
Thanks for your nice comments! - If you have any frames from a video that you can share, I'd love to be able to share them here. Please send then to gopro-overlay [at] time4tea.net . Of course, please don't feel obliged.

Currently there isn't a facility for max/min/avg of any metric. This is a good suggestion - I'll have a think about how this could be implemented.

**Reply by KosmoJoe** [reply]
Hi, thanks for your reply. I will check if I can share a video or at least an image :) 
Is it Ok if I make a suggestion for such an implementation? If so, I can try to find the time.
Bests

**Comment by KosmoJoe** [comment]
So I tried to understand the code a bit. My suggestion for inclusion of Min/Max Values would be:

Add in gopro-dashboard.py in the "main" under the processing Loop (with PoorTimer("processing").timing():) an addtional call to a function frame_meta.process(timeseries_process.calculate_minmax()). In this function, just iterate (as in calculate_odo()) through the time series and compute the min/max of "all" (needed) metrics. I would suggest to store them as the other metrics in the timeseries with a clear name which allows also to acces the value when only the metric is known (e.g. "odo.max" and "odo.min"). Storing this single value in the same way as the other metrics is a bit sloppy, however it allows on the one hand to use them as the other metrics in layouts and on the other hand excess them in every frame. I think also the function "def metric_accessor_from(name):" from layout_xml.py needs to be adapted to include the new metrics. Other needed changes I currently do not see, however, I have not implemented it ;)

In addtion, such an implementation would allow to use these "new" metrics in widgets: For example, one could use in the layout.xml file a statement such as
<component type="bar" width="850" height="80" metric="odo" units="meter" min="Min" max="Max"/>
 to indicate that the bar widget should use the min and max values of the metric "odo" for the min an max range of the bar itself. For this to work, we need to make the function "def create_bar()" in the class "class Widgets:" from layout_xml.py to call the metric_value() function for the "metric"+".max" when in the "max" attribute the string "Max" is found.

What do you think of these suggestion? If you think it is meaningful, I can also provide code.

**Reply by time4tea** [reply]
Hey, yes I think that would work exactly! I may have some time to implement this in the next couple of days.

I'm not worried at all about memory use etc. I'm only fixing performance and memory problems when they are an actual problem. So far that has worked OK. Storing a few extra values per frame is no big deal.

**Comment by WobwobRt** [comment]
Hey @time4tea great project! Already made an awesome dashboard for my road bike racing footage. 

Sorry if I’m hijacking, but I would like to add to the suggestion of @KosmoJoe the ability to display remaining distance. Is that doable?

**Reply by time4tea** [reply]
Hi - thanks for your interest. If you have an example frame that you don't mind sharing, it would be great to see. There is a category called Show and Tell: https://github.com/time4tea/gopro-dashboard-overlay/discussions/categories/show-and-tell

I guess this should be possible. There seem to be a few variants: distance to end of route (if this is available in fit/gpx), distance to end point (as crow flies), distance remaining along route. This last one is the most tricky I think, as distance estimation between points is actually the least accurate thing at the moment. We'd need to figure out what a good algorithm was to do that somewhat accurately.

**Reply by WobwobRt** [reply]
Hi @time4tea,

The .fit-files seem to contain the total distance (per lap). The accumulated distance could be used to subtract the ridden distance to get the remaining distance. I don't see this (total distance) value in .gpx-files. 

Not sure how accurate the total calculated ridden distance is compared to the total distance from the .fit-file, but in the .fit-file the current distance is also recorded:
![image](https://github.com/time4tea/gopro-dashboard-overlay/assets/140562858/b2cc77a0-2318-4d2b-8f33-ee73b3b9c613)

So, when using the a .fit-file, I guess it makes calculating the current and total distance quite easy as compared to the calculated current distance when using .gpx-files. 

Check the time and top right corners for what I have in mind (courtesy of NorCal Cycling):
![image](https://github.com/time4tea/gopro-dashboard-overlay/assets/140562858/dae23433-aac6-4a74-8a5b-b550fc3525e7)
![image](https://github.com/time4tea/gopro-dashboard-overlay/assets/140562858/ea0aa4b0-a0ca-4e7f-a1c8-cb47d6189ef4)

**Reply by WobwobRt** [reply]
Ah! Even easier, the total distance is also recorded in .fit-files:
![image](https://github.com/time4tea/gopro-dashboard-overlay/assets/140562858/541e4e1f-1484-4cd6-a89a-1e75f8280922)

**Reply by WobwobRt** [reply]
Might take a little time before I can get back on this. Currently don't have any material, but hoping to do some races outdoors soon again.

**Comment by time4tea** [comment]
I am looking into this.
